tests/python/pants_test/util/test_meta.py

```diff
@@ -242,3 +244,63 @@ class Concrete2(Abstract):
       def f(cls):
         return 'hello'
     self.assertEqual(Concrete2.f, 'hello')
+
+
+class FreezeAfterInitTest(unittest.TestCase):
+
+  def test_no_init(self) -> None:
+    @freeze_after_init
+    class Test:
+      pass
+
+    with self.assertRaises(FrozenInstanceError):
+      test = Test()
+      test.x = 1
+
+  def test_init_still_works(self):
+    @freeze_after_init
+    class Test:
+
+      def __init__(self, x: int) -> None:
+        self.x = x
+        self.y = "abc"
+
+    test = Test(x=0)
+    self.assertEqual(test.x, 0)
+    self.assertEqual(test.y, "abc")
+
+  def test_modify_preexisting_field_after_init(self) -> None:
+    @freeze_after_init
+    class Test:
+
+      def __init__(self, x: int) -> None:
+        self.x = x
+
+    with self.assertRaises(FrozenInstanceError):
+      test = Test(0)
+      test.x = 1
+
+  def test_add_new_field_after_init(self) -> None:
+    @freeze_after_init
+    class Test:
+
+      def __init__(self, x: int) -> None:
+        self.x: x
+
+    with self.assertRaises(FrozenInstanceError):
+      test = Test(0)
+      test.y = "abc"
+
+  def test_works_with_dataclass(self) -> None:
+    @freeze_after_init
+    @dataclass
```

Comment:
This is just a suggestion to make it more clear what's being tested, subjective.
```suggestion
    @dataclass(frozen=False)
```